### PR TITLE
DAOS-13711 client: several coverity silencing fixes

### PR DESCRIPTION
**Pull request by mchaarawi**
- Fix for coverity: 1449753 1449752 1449751 709506 1449749 1449748.
- Remove some uneeded allocations in array set size when adding 1 record.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by wangshilong — DISMISSED** [review]

File: src/client/array/dc_array.c [hunk 1]
```diff
@@ -2145,17 +2147,20 @@ check_record_cb(tse_task_t *task, void *data)
 
 	/** set memory location */
 	D_ALLOC(val, params->cell_size);
+	if (val == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
 	sgl->sg_nr = 1;
-	D_ALLOC_PTR(sgl->sg_iovs);
+	sgl->sg_iovs = &params->sg_iov;
 	d_iov_set(&sgl->sg_iovs[0], val, params->cell_size);
+	params->user_sgl_used = true;
```
Comment (liuxuezhao) [inline src/client/array/dc_array.c:2155]: just confirm, not very sure for the user_sgl_used flag usage, here it actually uses the params' internal inline sgl/sg_iov, it is not user API task's passed in sgl, right?

File: src/client/array/dc_array.c [hunk 2]
```diff
@@ -2287,30 +2299,33 @@ add_record(daos_handle_t oh, daos_handle_t th, struct set_size_props *props, d_l
...
-	/** set memory location */
+	/** set memory location - freed in free_set_size_cb (already registered) */
 	D_ALLOC(props->val, props->cell_size);
+	if (props->val == NULL)
+		D_GOTO(err, rc = -DER_NOMEM);
 	sgl->sg_nr = 1;
-	D_ALLOC_PTR(sgl->sg_iovs);
+	sgl->sg_iovs = &params->sg_iov;
 	d_iov_set(&sgl->sg_iovs[0], props->val, props->cell_size);
+	params->user_sgl_used = true;
```
Comment (liuxuezhao) [inline src/client/array/dc_array.c:2312]: same, the user_sgl_used flag looks a little confuse to me, but i am not familiar with those detailed code.

**Review by liuxuezhao — APPROVED** [review]

File: src/client/array/dc_array.c (hunk 1 above)
Comment (mchaarawi) [reply src/client/array/dc_array.c:2155]: it is a little confusing. in this case, the calling function here is the user and we are using a preallocated sgl from the struct. it is not needed to allocate 1 here. we set this flag to true so to not free the iov it in the free function cb and we just free the whole args struct.
before we were just allocating and freeing an iov of size 1 which is not ideal (required for async operations to be on the heap not the stack).

File: src/client/array/dc_array.c (hunk 2 above)
Comment (mchaarawi) [reply src/client/array/dc_array.c:2312]: same as above.

**Review by liuxuezhao — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Pull request merged**